smart-contract-feature-development.md

```diff
@@ -0,0 +1,94 @@
+# Purpose
+
+The purpose of this document is to provide guidance on how develop smart contracts without the need for feature branches.
+
+# Summary
+
+Feature branches result in a lot of effort spent keeping the feature branch up to date. We want to have modular code
+that enables us to easily compose features together such that we can get code mainlined early and toggle it off when
+running in production.
+
+# Problem Statement + Context
+
+We want to move away from a world with long lived feature branches.
+To do so, we need to have some flexibility when it comes to composition of features.
+This is easy in off-chain code as the cost is minimal for toggling off certain features but still having them in the source code.
+This is not the case for Solidity. This document exists to outline how to think about composition of features for Solidity code.
+When the feature is ready to be shipped, the feature specific contract should be merged into the base contract.
+
+# Alternatives Considered
+
+## Copy and Paste Implementation
+
+This was the solution for the fault proofs project, which worked well until the custom gas token feature was introduced. Now the contracts
+are in a state such that the custom gas token feature needs to be ported to the `OptimismPortal2` before the `OptimismPortal` and the
+`OptimismPortal2` can be merged. This may be a good solution for security critical code as we do not rely on the correctness
+of solidity inheritance and all of the logic lives in a single contract, making it easier to audit.
+
+# Proposed Solution
+
+Below includes general guidelines for how to develop solidity feature contracts that are meant to be merged into mainline at
+some point in the future.
+
+### Composition through Inheritance
+
+Solidity allows contacts to [inherit](https://docs.soliditylang.org/en/v0.8.17/contracts.html#inheritance) other contracts.
+The functionality of the contract can then be extended or overridden.
+For in development features that modify core contracts already running in production,
+a new contract should be created that inherits the core contract and all new functionality should be added to the new contract.
+The deploy script should be able to parse some sort of config and optionally deploy the feature enabled version of the contract.
+
+### Handling Inheritance
```

Comment:
For fault proofs, it would have been possible to follow this approach by inheriting the contract and adding new functions. Functions that were removed in the new portal would have to be overridden and revert. You cannot subtract interfaces under this approach, which isn't ideal. I think that if you need to subtract interfaces, then its better dealt with composability of smaller interfaces, which is in the family of the approach mentioned. This would look like composing `OptimismPortal` from 2 contracts, one is all deposits and one is all withdrawals, then for `OptimismPortal2` implement a new withdrawal contract and compose it with the same deposit contract.

For interop, we are only adding interfaces, so we do not have this problem.

I suppose this is a good callout of how to think about the need to add or remove interfaces